mm2src/lp_ordermatch/simple_market_maker.rs

```diff
@@ -0,0 +1,530 @@
+    let time_diff = rates.retrieve_elapsed_times();
+    let elapsed = match time_diff.elapsed() {
+        Ok(elapsed) => elapsed.as_secs_f64(),
+        Err(_) => return MmError::err(OrderProcessingError::PriceElapsedValidityUntreatable),
+    };
+    let elapsed_validity = cfg.price_elapsed_validity.unwrap_or(300.0);
+
+    if elapsed > elapsed_validity {
+        warn!(
+            "last updated price timestamp elapsed {} is more than the elapsed validity {} - skipping for {}",
+            elapsed, elapsed_validity, key_trade_pair,
+        );
+        cancel_functor().await;
+        return MmError::err(OrderProcessingError::PriceElapsedValidityExpired);
+    }
+    info!("elapsed since last price update: {} secs", elapsed);
+    Ok(true)
+}
+
+async fn update_single_order(
+    cfg: SimpleCoinMarketMakerCfg,
+    uuid: Uuid,
+    _order: MakerOrder,
+    key_trade_pair: String,
+    ctx: &MmArc,
+) -> OrderProcessingResult {
+    info!("need to update order: {} of {} - cfg: {}", uuid, key_trade_pair, cfg);
+    let simple_market_maker_bot_ctx = TradingBotContext::from_ctx(ctx).unwrap();
+    let registry = simple_market_maker_bot_ctx.price_tickers_registry.lock().await;
+    let rates = registry.get_cex_rates(cfg.base.clone(), cfg.rel.clone());
+    drop(registry);
+    checks_order_prerequisites(ctx, &rates, &cfg, key_trade_pair.clone(), Option::from(uuid)).await?;
+    Ok(true)
+}
+
+async fn create_single_order(
+    cfg: SimpleCoinMarketMakerCfg,
+    key_trade_pair: String,
+    ctx: &MmArc,
+) -> OrderProcessingResult {
+    info!("need to create order for: {} - cfg: {}", key_trade_pair, cfg);
+    let simple_market_maker_bot_ctx = TradingBotContext::from_ctx(ctx).unwrap();
+    let registry = simple_market_maker_bot_ctx.price_tickers_registry.lock().await;
+    let rates = registry.get_cex_rates(cfg.base.clone(), cfg.rel.clone());
+    drop(registry);
+
+    checks_order_prerequisites(ctx, &rates, &cfg, key_trade_pair.clone(), None).await?;
+
+    let base_balance = coin_find_and_checks(cfg.base.clone(), key_trade_pair.clone(), true, ctx).await?;
+    coin_find_and_checks(cfg.rel.clone(), key_trade_pair.clone(), false, ctx).await?;
+
+    info!("balance for {} is {}", cfg.base, base_balance);
+
+    let mut calculated_price = rates.price * cfg.spread;
+    info!("calculated price is: {}", calculated_price);
+    if cfg.check_last_bidirectional_trade_thresh_hold.unwrap_or(false) {
+        vwap_apply(&mut calculated_price).await;
+    }
+
+    let volume = match cfg.balance_percent {
+        Some(balance_percent) => balance_percent * base_balance.clone(),
+        None => MmNumber::default(),
+    };
+
+    let min_vol: Option<MmNumber> = match cfg.min_volume {
+        Some(min_volume) => {
+            if cfg.max.unwrap_or(false) {
+                Option::from(min_volume * base_balance.clone())
+            } else {
+                Option::from(min_volume * volume.clone())
```

Comment:
`from` is used to convert the types from each other. `Some(...)` is a proper way here.